### ISSUE TITLE
`SeqView`: changing operations on Sequences to act as views until necessary 

### ISSUE THREAD

**Issue by KatherineCaley** [cogent3/cogent3#1285]
### Context:
 We need operations on Sequences to behave as views until it's essential the underlying data is operated on. The first challenge is to implement consistent slice transactions. 

#### Current implementation:
The `Sequence` object has a `._seq` attribute, previously this is the literal sequence string, and now this is wrapped in a `SeqView` object. 

Here is the core of the `SeqView` class, which handles the slice transactions:

``` python3
class SeqView:
    __slots__ = ("seq", "start", "stop", "step", "replacements")

    def __init__(self, seq, *, start: int = None, stop: int = None, step: int = None):
        self.seq = seq
        self.start = start or 0
        self.stop = len(seq) if stop is None else stop
        self.step = step or 1
        self.replacements = []

    def __getitem__(self, slice: slice):
        step = self.step if slice.step is None else slice.step
        start = slice.start if slice.start is not None else 0 if step > 0 else -1
        stop = slice.stop if slice.stop is not None else len(self) if step > 0 else -len(self) -1

        return self.__class__(
            self.seq,
            start=self.start + start * abs(self.step),
            stop=min(self.stop, self.start + stop * abs(self.step)),
            step=self.step * step,
        )

    @property
    def value(self):
        start, stop, step = self.start, self.stop, self.step

        value = self.seq[start:stop:step]
        for old, new in self.replacements:
            value = value.replace(old, new, -1)
        return value

    def replace(self, old, new):
        if len(old) == len(new):
            self.replacements.append((old, new))
            return self
        else:
            raise ValueError("old and new must be the same length")

    def __len__(self):
        return abs((self.start - self.stop) // abs(self.step))

```

### To do/to think about:
- get test suite passing with these changes 
- think about how to measure the performance of this against the original implementation in terms of both memory and speed.
- How to determine equality between two SeqView objects, just the value of the slice?
- How to handle replacements that are going to change the length of the underlying data, probably using Map.

**Comment by KatherineCaley** [comment]
If we wanted a `SeqView` object to show multiple slices of its underlying sequence, what would be the best way? Is it a `Map` object, or a series of start/stop/step that are applied sequently? 

For example, the method on `Sequence` called`parse_out_gaps` where you want to get the sequence without gap characters, but without losing the relationship between the positions in the original gapped sequence and the corresponding positions in the gapless sequence. 

```python 
def parse_out_gaps(self):
        gapless = []
        segments = []
        nongap = re.compile(f"([^{re.escape('-')}]+)")
        for match in nongap.finditer(
            self._seq.value
        ):
            segments.append(match.span())
            gapless.append(match.group())
        map = Map(segments, parent_length=len(self)).inverse()
        seq = self.__class__(
            "".join(gapless), name=self.get_name(), info=self.info, preserve_case=True
        )
        if self.annotations:
            seq.annotations = [a.remapped_to(seq, map) for a in self.annotations]
        return (map, seq)
```

**Comment by GavinHuttley** [comment]
I suggest we add a `__str__` method which returns `self.view`.

In the `view()` method, rather than
```python
        for old, new in self.replacements:
            value = value.replace(old, new, -1)
```
We should make a translation table, it will be **much** faster
```python
char_map = str.maketrans(*list(zip(*self.replacements)))
return value.translate(char_map)
```

**Comment by KatherineCaley** [comment]
@GavinHuttley, design choice. Should the view be accessed by `.view` attribute or `.value`. 

The original code uses `.value` but i can see the merit in it being obvious to a user that they are not accessing the underlying sequence object.

**Comment by GavinHuttley** [comment]
Typo on my part, it should be value.

**Comment by GavinHuttley** [comment]
fixed in e330639
